GPGPU Support

mastercoms opened the pull request:
WIP branch for implementing GPGPU support within Glowstone. In this branch, it is used for noise generation, but can be easily applied to other systems after this is merged.

The scope of this PR could change to remove the noise gen modification for quicker merging of OpenCL support, since the OpenCL backend is sufficiently completed, however, it could have better support of OpenCL code. Perhaps a folder dedicated for third party code that could be used by plugins.

turt2live commented:
Closing because the partner PR is closed and this will not compile without it. If the partner PR is reopened, or this can be addressed in the future, it may be reopened.

mastercoms committed:
Add JogAmp JOCL dependency

mastercoms committed:
Add JogAmp JOCL dependency

mastercoms committed:
Initial work with JOCL

mastercoms committed:
Initial work with JOCL

mastercoms committed:
Some more work on opencl

mastercoms committed:
Some more work on opencl

mastercoms committed:
Implement lerp, test it in final noise gen

mastercoms committed:
Implement lerp, test it in final noise gen

mastercoms committed:
More advanced config and backend for OpenCL support

mastercoms committed:
More advanced config and backend for OpenCL support

mastercoms committed:
Cache OpenCL context and queue

mastercoms committed:
Cache OpenCL context and queue

mastercoms committed:
Align method calls

mastercoms committed:
Align method calls

mastercoms committed:
Directly use system classloader resource stream getter

mastercoms committed:
Directly use system classloader resource stream getter

mastercoms committed:
Fix imports

I had an urge to do this, but i did it on my phone while doing something else

cleared my mind but did not write working code.

mastercoms committed:
Fix imports

I had an urge to do this, but i did it on my phone while doing something else

cleared my mind but did not write working code.

mastercoms committed:
Assign isCLApplicable to relevant value

mastercoms committed:
Assign isCLApplicable to relevant value

mastercoms committed:
Redo branching due to changed boolean

mastercoms committed:
Redo branching due to changed boolean

Deamon5550 commented:
Do you have any benchmarks for this with the noise gen?

mastercoms commented:
@Deamon5550 Yes, but right now it is not focused on speed, just on a working implementation. I am working on optimizing it now that it is working. Right now, it takes about 2.5 minutes to generate a world compared to 10 seconds with the original implementation, but this is because of unoptimized loops and bad memory management. For the OpenCL implementation to be faster, all of the noise generation needs to be implemented in OpenCL, so that's what I'm working on right now. Once I have that pushed, I'll publish benchmarks.

mastercoms committed:
Lower requirement to OpenCL 1.2 for testing, fix device support checking

mastercoms committed:
Lower requirement to OpenCL 1.2 for testing, fix device support checking

mastercoms committed:
Up requirement to OpenCL 2.0

mastercoms committed:
Up requirement to OpenCL 2.0

mastercoms committed:
cl_khr_fp64 is core in 1.2

mastercoms committed:
cl_khr_fp64 is core in 1.2

mastercoms committed:
Fix device detection, use correct resource loader

mastercoms committed:
Fix device detection, use correct resource loader

mastercoms committed:
Fix null InputStream

mastercoms committed:
Fix null InputStream

mastercoms committed:
Add new debug info, temp switch for GPGPU in noise gen, activate cl_khr_fp64 in cl code

mastercoms committed:
Add new debug info, temp switch for GPGPU in noise gen, activate cl_khr_fp64 in cl code

mastercoms committed:
Separate Intel graphics, because they report flops that aren't comparable to dGPU flops

mastercoms committed:
Separate Intel graphics, because they report flops that aren't comparable to dGPU flops

mastercoms committed:
Fix some variable errors

mastercoms committed:
Fix some variable errors

mastercoms committed:
Better OpenCL noise gen

mastercoms committed:
Better OpenCL noise gen

mastercoms committed:
Merge from master

mastercoms committed:
Merge branch 'master' into gpu3

mastercoms committed:
Be less strict with opencl detection

mastercoms commented:
Most work from this PR has been integrated into master with https://github.com/GlowstoneMC/Glowstone/commit/2b731f3a1d2709e15f5b9bcb9f63644be12a8953. This commit does not include any usage of OpenCL, unlike this PR. Usage of OpenCL will be done later in master.
